### PR TITLE
test(endpoints_tests): add tests for app

### PR DESCRIPTION
**Pull request by AlexMomot-717**

### CODE REVIEW

File: .github/workflows/hexlet-check.yml
Comment (svntmr) [inline .github/workflows/hexlet-check.yml:1]: should be moved to a separate commit with an explanation of why we do that

File: .pre-commit-config.yaml
Comment (svntmr) [inline .pre-commit-config.yaml:1]: should be included in the same commit as the one where `hexlet-check.yml` is deleted

File: README.md
Comment (svntmr) [inline README.md:1]: should be included in the same commit as the one where `hexlet-check.yml` is deleted

File: tests/conftest.py [hunk 1]
```diff
@@ -1,9 +1,27 @@
...
+        }
+    )
+    test_app.secret_key = "dummy-secret-key"
+
+    return test_app
+
+
+@pytest.fixture
+def client(app):
+    return app.test_client()
```
Comment (svntmr) [inline tests/conftest.py]: as we're not doing any complicated setup in the `app` fixture, we can unify them into one. wdyt?

File: tests/test_app.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    route = "/urls"
+
+    # when
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    url_id = get_url_by_name(url_name)
+    assert url_id == 1
+    assert response.request.path == "/urls/1"
+    assert "Страница успешно добавлена" in response.text
```
Comment (svntmr) [inline tests/test_app.py]: it's better to do response checks first and then check the side effects (DB entry creation)

File: tests/test_app.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+from page_analyzer.utils.db import (
+    create_check,
+    create_url,
+    get_url_by_name,
+    get_url_checks,
+    get_url_data,
+)
+
+
+def test_index(client):
```
Comment (svntmr) [inline tests/test_app.py]: please, add type hints

File: tests/test_app.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    # given
+    route = "/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 200
+    assert "Бесплатно проверяйте сайты на SEO-пригодность" in response.text
```
Comment (svntmr) [inline tests/test_app.py]: wdyt, is it enough to know that we return the stable and reproducible page content?

File: tests/test_app.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 200
+    assert b"<td>https://www.example.com</td>" in response.data
+
+
+@freeze_time("2024-05-01")
+def test_show_url_checks(test_database, client):
```
Comment (svntmr) [inline tests/test_app.py]:
```suggestion
from flask.testing import FlaskClient

def test_show_url_with_checks(test_database: None, client: FlaskClient) -> None:
```

File: tests/test_app.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        create_check(url_data1)
+
+    # when
+    response = client.get(route)
+
+    # then
+    expected_text_elements = ["1", "2", url_name1, url_name2, "200", "2024-05-01"]
+    assert response.status_code == 200
+    for text in expected_text_elements:
+        assert text in response.text
```
Comment (svntmr) [inline tests/test_app.py]: too complicated to understand what's checked

**Review by svntmr — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 7]
```diff
@@ -1,9 +1,21 @@
 import os
 
 import pytest
+from page_analyzer.app import app as test_app
 from page_analyzer.utils.db import connect_db
 
 
+@pytest.fixture
+def client():
```
Comment (svntmr) [inline tests/conftest.py]: let's add type hints!

File: tests/fixtures/expected_htmls.py
Comment (svntmr) [inline tests/fixtures/expected_htmls.py:1]: wdyt, does it make sense to move values to actual HTML files? and load their content here

File: tests/test_app.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+def test_index(client: FlaskClient) -> None:
+    # given
+    route = "/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 200
+    assert EXPECTED_INDEX_HTML in response.text
```
Comment (svntmr) [inline tests/test_app.py]: why not `==`?

File: tests/test_app.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    # given
+    url_name = "https://www.example.com"
+    create_url(url_name)
+    route = "/urls"
+
+    # when
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    assert "Страница уже существует" in response.text
```
Comment (svntmr) [inline tests/test_app.py]: let's add html for this page

File: tests/test_app.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    # given
+    url_id = 0
+    route = f"/urls/{url_id}"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 200
+    assert "Страница не найдена" in response.text
```
Comment (svntmr) [inline tests/test_app.py]: let's add html content here

File: tests/test_app.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            "id": 1,
+            "url_id": 1,
+            "status_code": 200,
+            "h1": "some header",
+            "title": "some title",
+            "description": "some description",
+            "created_at": datetime.date(2024, 5, 1),
+        }
+    ]
+    assert response.status_code == 200
```
Comment (svntmr) [inline tests/test_app.py]: let's check response first and side effect later :)

File: tests/test_app.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    # given
+    url_id = 99
+    route = f"/urls/{url_id}/checks"
+
+    # when
+    response = client.post(route, follow_redirects=True)
+
+    # then
+    assert response.status_code == 200
+    assert "Страница не найдена" in response.text
```
Comment (svntmr) [inline tests/test_app.py]: let's add html here

**Review by svntmr — CHANGES_REQUESTED** [review]

File: tests/fixtures/expected_htmls.py
Comment (AlexMomot-717) [reply tests/fixtures/expected_htmls.py:1]: it will work for clear index.html and page_not_found.html only, cause those don't depend on context from endpoints

Comment (svntmr) [reply tests/fixtures/expected_htmls.py:1]: what context?

Comment (AlexMomot-717) [reply tests/fixtures/expected_htmls.py:1]: I mean (context is about django :)) url_data, checks_data etc

File: tests/test_app.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+    get_url_by_name,
+    get_url_checks,
+    get_url_data,
+)
+from tests.paths import resolve_path
+
+
+def expected_html_data(html_file):
+    with open(html_file, "r", encoding="utf-8") as f:
+        return f.read()
```
Comment (svntmr) [inline tests/test_app.py]: you can user `resolve_path` and `Path.read_bytes()` instead

File: tests/test_app.py [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    # given
+    route = "/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    html_file = str(resolve_path("fixtures/expected_index.html"))
+    assert expected_html_data(html_file) == response.text
+    assert response.status_code == 200
```
Comment (svntmr) [inline tests/test_app.py]:
```suggestion
    expected_html = resolve_path("fixtures/expected_index.html")
    assert response.text == expected_html.read_bytes()
    assert response.status_code == 200
```

**Review by svntmr — CHANGES_REQUESTED** [review]
close!

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -45,14 +47,15 @@ repos:
...
-        entry: mypy page_analyzer --config-file pyproject.toml
+        entry: mypy page_analyzer tests --config-file pyproject.toml
         pass_filenames: false # not needed, as mypy should be executed on the whole codebase
         always_run: true
         types: [python, pyi]
   - repo: https://github.com/pycqa/flake8
     rev: 7.0.0
     hooks:
       - id: flake8
+        exclude: tests/fixtures
```
Comment (svntmr) [inline .pre-commit-config.yaml]: I guess we can remove this exclude now as fixtures are HTML files now - flake shouldn't care about them

File: page_analyzer/utils/db.py [hunk 16]
```diff
@@ -89,7 +89,9 @@ def get_urls() -> List[Dict[str, Any]]:
     return actual_urls_data
 
 
-def create_check(url_data: Dict[str, Any]) -> bool:
+def create_check(url_data: Dict[str, Union[int, str]] | None) -> bool:
```
Comment (svntmr) [inline page_analyzer/utils/db.py]: In modern versions of Python ([since 3.10](https://docs.python.org/3/library/typing.html#typing.Union)) `Union` can be replaced with `|`

```suggestion
def create_check(url_data: Dict[str, int | str] | None) -> bool:
```

File: page_analyzer/utils/db.py [hunk 17]
```diff
@@ -89,7 +89,9 @@ def get_urls() -> List[Dict[str, Any]]:
...
+    if url_data is None:
+        return False
```
Comment (svntmr) [inline page_analyzer/utils/db.py]: will this function ever be called with `None` as input?

File: pyproject.toml [hunk 18]
```diff
@@ -50,10 +50,12 @@ line_length = 79
 [tool.mypy]
 python_version = 3.11
 strict = true
+ignore_missing_imports = true
+disallow_untyped_decorators = false
+strict_equality = false
```
Comment (svntmr) [inline pyproject.toml]: what was the reason for adding these settings?

File: tests/conftest.py [hunk 19]
```diff
@@ -1,11 +1,25 @@
...
+        {
+            "TESTING": True,
+        }
+    )
+    test_app.secret_key = "dummy-secret-key"
+    return test_app.test_client()
+
+
+@pytest.fixture
+def test_database() -> Generator[Any, None, None]:
```
Comment (svntmr) [inline tests/conftest.py]: as `Generator` is typed using `Generator[YieldType, SendType, ReturnType]` ([context](https://docs.python.org/3/library/typing.html#typing.Generator)), this function should be defined in this way:
```suggestion
def test_database() -> Generator[None, None, None]:
```

File: tests/fixtures/__init__.py
Comment (svntmr) [inline tests/fixtures/__init__.py:1]: this folder is not a python package, this file can be removed

File: tests/test_app.py [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+    route = "/urls"
+
+    # when
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_add.html")
+    assert response.data == expected_html.read_bytes()
+    url_id = get_url_by_name(url_name)
+    assert url_id == 1
```
Comment (svntmr) [inline tests/test_app.py]: [optional] can be inlined

File: tests/test_app.py [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+    url_name = "https://www.example.com"
+    create_url(url_name)
+    route = "/urls"
+
+    # when
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_exists.html")
+    assert response.data == expected_html.read_bytes()
```
Comment (svntmr) [inline tests/test_app.py:57]: what is the response code in this case?

File: tests/test_app.py [hunk 22]
```diff
@@ -0,0 +1,242 @@
...
+    # given
+    url_name = "https://www.example.com"
+    route = "/urls"
+
+    # when
+    response = client.post(route, data={"url": url_name}, follow_redirects=True)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_add.html")
+    assert response.data == expected_html.read_bytes()
```
Comment (svntmr) [inline tests/test_app.py:40]: what's the response code?

File: tests/test_app.py [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+@freeze_time("2024-05-01")
+def test_add_url_check_error(client: FlaskClient, test_database: None) -> None:
+    # given
+    url_name = "https://www.example1.com"
+    url_id = create_url(url_name)
+    route = f"/urls/{url_id}/checks"
+
+    # when
+    with patch("page_analyzer.utils.db.check_url", return_value=None):
+        response = client.post(route, follow_redirects=True)
```
Comment (svntmr) [inline tests/test_app.py:222]: should this part raise instead? test function says `error`

**Review by svntmr — CHANGES_REQUESTED** [review]

File: tests/test_app.py (hunk 23 above)
Comment (AlexMomot-717) [reply tests/test_app.py:222]: I consider this side effect could be raised inside check_url test.
we get None here as a result of check_url raise error.

Comment (svntmr) [reply tests/test_app.py:222]: let's rename this test to smth like "...check_fails"

File: pyproject.toml (hunk 18 above)
Comment (AlexMomot-717) [reply pyproject.toml]: image

Comment (AlexMomot-717) [reply pyproject.toml]: tests/conftest.py:21: error: Untyped decorator makes function "test_database" untyped

tests/utils/test_db.py:25: error: Non-overlapping equality check (left operand type: "Optional[Dict[str, Union[int, str]]]", right operand type: "List[object]")
tests/utils/test_db.py:39: error: Non-overlapping equality check (left operand type: "Optional[Dict[str, Union[int, str]]]", right operand type: "List[object]")
tests/utils/test_db.py:88: error: Non-overlapping equality check (left operand type: "List[Dict[str, Any]]", right operand type: "List[List[object]]")

Comment (svntmr) [reply pyproject.toml]: first issues can be resolved by a couple of isinstance asserts

File: page_analyzer/utils/db.py (hunk 17 above)
Comment (AlexMomot-717) [reply page_analyzer/utils/db.py]: you're right:  
get_url_data(url_id) can return  None,
but function add_url_check catches that None:

def add_url_check(id: int) -> Any:
    url_data = get_url_data(id)
    if not url_data:
        return render_template("page_not_found.html")
    check_context = create_check(url_data)

However mypy ignores this fact and requires optional None

Comment (svntmr) [reply page_analyzer/utils/db.py]: mypy won't know about 'None' if you remove it from the arguments typing

File: tests/test_app.py (hunk 23 above)
Comment (svntmr) [reply tests/test_app.py:222]: @AlexMomot-717 bump

**Review by svntmr — APPROVED** [review]
one small comment from the previous review
